esmvalcore/_main.py

```diff
@@ -68,14 +68,13 @@ def parse_resume(resume, recipe):
     return resume
 
 
-def process_recipe(recipe_file, config_user):
+def process_recipe(recipe_file: Path, session):
```

Comment:
Adding type hints for all functions is rather beyond the scope of what I intended to do with this pull request. It is a lot of work to get that right and introduces many additional changes because the mypy checks will not pass without them.

Reply:
When you try to add type hints, you very quickly run into circular imports. It would be good to restructure the code as much as possible so that doesn't happen, but that is quite a bit of work so I think it would be best to add them when we can until we have them everywhere. Another pattern that we commonly use that mypy doesn't like is reusing the same variable name for something that is the same but has a different type.

In the case of esmvalcore/_main.py, we only want to import things when running the code to keep the command line tool responsive, so we would need to add special imports for type checking in this case.